packages/hardhat-core/src/builtin-tasks/test.ts

```diff
@@ -43,24 +44,59 @@ subtask(TASK_TEST_GET_TEST_FILES)
 subtask(TASK_TEST_SETUP_TEST_ENVIRONMENT, async () => {});
 
 subtask(TASK_TEST_RUN_MOCHA_TESTS)
+  .addFlag("parallel", "Run tests in parallel")
+  .addFlag("bail", "Stop running tests after the first test failure")
+  .addOptionalParam("grep", "Only run tests matching the given regexp")
   .addOptionalVariadicPositionalParam(
     "testFiles",
     "An optional list of files to test",
     []
   )
-  .setAction(async ({ testFiles }: { testFiles: string[] }, { config }) => {
-    const { default: Mocha } = await import("mocha");
-    const mocha = new Mocha(config.mocha);
-    testFiles.forEach((file) => mocha.addFile(file));
+  .setAction(
+    async (
+      {
+        bail,
+        parallel,
+        testFiles,
+        grep,
+      }: {
+        bail: boolean;
+        parallel: boolean;
+        testFiles: string[];
+        grep?: string;
+      },
+      { config }
+    ) => {
+      const { default: Mocha } = await import("mocha");
 
-    const testFailures = await new Promise<number>((resolve) => {
-      mocha.run(resolve);
-    });
+      const mochaConfig: MochaOptions = {
+        ...config.mocha,
+        bail,
+        grep,
```

Comment:
Should/do we allow these to be configured in `config.mocha` as well? If so, this would overwrite them (i.e. if they were set in `config.mocha` but not passed as params to the subtask).

Reply:
Oh, I hadn't noticed that, thanks for pointing it out. I actually think that would be the correct behavior though? Normally CLI flags override what's in the config (for example, that's what Prettier does by default). But it's good to know that this is the case. I will add a comment explaining that this is intentional.

Reply:
Wait I see what you mean. For the `grep` flag this works correctly, but `bail` has a default value, and so it will always override whatever is in the config. I'll fix this, thank you.

Reply:
Great catch! I was also tricked by this

Reply:
Fixed.